app/components/channel_link/index.js

```diff
@@ -3,26 +3,64 @@
 
 import {connect} from 'react-redux';
 import {bindActionCreators} from 'redux';
+import {createSelector} from 'reselect';
 
-import {getChannelsNameMapInCurrentTeam} from 'mattermost-redux/selectors/entities/channels';
+import {General} from 'mattermost-redux/constants';
+import {
+    getChannelsNameMapInCurrentTeam,
+    getChannelsInCurrentTeam,
+    getMyChannelMemberships,
+} from 'mattermost-redux/selectors/entities/channels';
+import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
+import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';
+import {joinChannel, searchChannels} from 'mattermost-redux/actions/channels';
 
 import {handleSelectChannel, setChannelDisplayName} from 'app/actions/views/channel';
 
 import ChannelLink from './channel_link';
 
-function mapStateToProps(state) {
-    return {
-        channelsByName: getChannelsNameMapInCurrentTeam(state),
+function makeGetJoinableChannel() {
+    return createSelector(
+        getChannelsInCurrentTeam,
+        getMyChannelMemberships,
+        (state, props) => props && props.channelMentions,
+        (channels, myMembers, channelMentions) => {
+            const displayname = channelMentions && Object.values(channelMentions)[0].display_name;
+            const joinableChannels = channels.filter((c) => {
+                return (!myMembers[c.id] && c.type === General.OPEN_CHANNEL && displayname && displayname === c.display_name);
+            });
+
+            if (joinableChannels.length === 1) {
+                return joinableChannels[0];
+            }
+
+            return null;
+        },
+    );
+}
+
+function makeMapStateToProps() {
+    const getJoinableChannel = makeGetJoinableChannel();
+
+    return function mapStateToProps(state, ownProps) {
+        return {
+            channelsByName: getChannelsNameMapInCurrentTeam(state),
+            currentTeamId: getCurrentTeamId(state),
+            currentUserId: getCurrentUserId(state),
+            joinableChannel: getJoinableChannel(state, ownProps),
```

Comment:
Why do we need to pass the `joinableChannel` separately? Couldn't we just find it in `channelsByName`?

If we need to determine whether or not you are a member of the channel before showing it, we could factor out that logic into an action creator so that we'd be able to check the `myMembers` when the user taps on the link instead of doing it before